### ISSUE TITLE
Incorporating flash-attention2 [SOLVED] and subsequent testing [ONGOING]

### ISSUE THREAD

**Issue by BBC-Esq** [SYSTRAN/faster-whisper#598]
Hello all.  Just thought I'd post a question about Flash Attention 2 here:

[https://github.com/Dao-AILab/flash-attention](https://github.com/Dao-AILab/flash-attention)

Apparently it's making big waves and seems seems very powerful.  Does anyone plan on seeing if it's something that could be included?

**Comment by phineas-pta** [comment]
not faster whisper issue

require a lot of new c++ code into ctranslate2, so highly doubt devs have time

**Comment by blackpolarz** [comment]
Before talking about adopting flash attention, it is necessary to understand what flash attention v2 does and how it helps. 
Flash attention is an library which reorganises the workflow, parallelise workflow and reducing the number of shifts in memory. This requires the implementer to have sufficient knowledge about the memory management of the gpu and even the original implementer of flash attention v2 only managed to make it compatible for certain gpus as of today. 
Rather than focusing on making flash attention library compatible with ctranslate2 which may or may not help with the speedup, it would make more sense to improve ctranslate2 further.

**Comment by BBC-Esq** [comment]
> ctranslate2 is backend of faster-whisper, all heavy computations are done in ctranslate2
> 
> same for pytorch is backend of openai-whisper

Thanks for clarifying...so basically, any improvements to the core functionality requires changing the code for ctranslate2 then...so in essence it's not a true faster-whisper issue like you state?  It's a little confusing to a non-programmer by trade like myself.  I look at the [website here](https://opennmt.net/CTranslate2/python/overview.html) and it seems to be a Python (API)...if that's the correct term?  But then faster-whisper is what, a "wrapper" to abstract some of the functionality, to make it easier to program in Python and use ctranslate2 under the hood?  Thus, does faster-whisper actually add any functionality/features or does it merely provide this abstraction layer to make programs easier with Python that use ctranslate2?

I'd definitely like to direct my suggestions to the appropriate repositories.  If it's a feature/functionality that requires a change to ctranslate2, I can submit an issue there...but if it's something that faster-whisper adds, I can definitely submit an issue here instead.  Thanks for your understanding to an in-experienced programmer/hobbyist.

**Comment by BBC-Esq** [comment]
Got it!  So I hope ya'll will forgive me if I accidentally make an issue here that should be directed at ctranslate2, but I'll try to divy up any future issues to the collaborators who actually would be responsible for any changes.  :-)

In theory, I could write a program to transcribe audio with the code (and Python API) located at the ctranslate2 repository that I linked...or I could use faster-whisper to do the same thing easier (plus have some additional functionality/features I'm not yet clear on)...And finally, I can use ctranslate2 via Langchain's implementation [like here](https://python.langchain.com/docs/integrations/llms/ctranslate2), is that correct?

And there's yet another implementation [that is here](https://michaelfeil.eu/hf-hub-ctranslate2/), if you're familiar?

So basically, ctranslate2 is being used under the hood but there's "abstraction" layers or "wrappers" or "APIs" (what's the proper term) that various people implement to use ctranslate2...but with any core functionality (like Flash Attention 2), I need to direct my issues to that repository, is that correct?

**Comment by junchen6072** [comment]
there is some discussion here on ctranslate2:
https://github.com/OpenNMT/CTranslate2/issues/1300

**Comment by trungkienbkhn** [comment]
@jet082 , It's supported from version 1.0.2 and is related to this [PR](https://github.com/SYSTRAN/faster-whisper/pull/807).
You can try this implementation:
```python
from faster_whisper import WhisperModel
model = WhisperModel('large-v3', device="cuda", flash_attention=True)
```
Ctranslate2 docs: https://opennmt.net/CTranslate2/python/ctranslate2.models.Whisper.html

**Comment by twicer-is-coder** [comment]
@jet082 After enabling flash_attention, how much speed boost did you noticed?

**Comment by twicer-is-coder** [comment]
> > @jet082 , It's supported from version 1.0.2 and is related to this [PR](https://github.com/SYSTRAN/faster-whisper/pull/807). This PR allows adding new model args from ctranslate2 to the FW, even if they have not been defined in the FW. You can try this implementation:
> > ```python
> > from faster_whisper import WhisperModel
> > model = WhisperModel('large-v3', device="cuda", flash_attention=True)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Ctranslate2 docs: https://opennmt.net/CTranslate2/python/ctranslate2.models.Whisper.html
> 
> My apologies, I was a version behind - this works great!

Can you share the performance improvement on Flash Attention?

**Comment by AvivSham** [comment]
Regarding benchmarking faster-whisper with FA - I did not witness any speed boost when loading the faster whisper model with FA. I shared the full details including code snip in this issue that originated in ctranslate2 repo https://github.com/OpenNMT/CTranslate2/issues/1300.

In short, when transcribing ~50 sec audio with the default faster whisper params (beam equals 5) there is no difference if you use FA or not the inference timed measured were almost identical. 
Does anyone have noticed the same behavior? if not what am I doing wrong? 

Thanks in advance,

@Purfview @BBC-Esq

**Comment by twicer-is-coder** [comment]
> Regarding benchmarking faster-whisper with FA - I did not witness any speed boost when loading the faster whisper model with FA. I shared the full details including code snip in this issue that originated in ctranslate2 repo [OpenNMT/CTranslate2#1300](https://github.com/OpenNMT/CTranslate2/issues/1300).
> 
> In short, when transcribing ~50 sec audio with the default faster whisper params (beam equals 5) there is no difference if you use FA or not the inference timed measured were almost identical. Does anyone have noticed the same behavior? if not what am I doing wrong?
> 
> Thanks in advance,
> 
> @Purfview @BBC-Esq

Yes, I didn't notice any difference in speed  too when FA enabled.

**Comment by AvivSham** [comment]
@Napuh 
maybe you can help with this?

**Comment by BBC-Esq** [comment]
> @Napuh maybe you can help with this?

BTW, I just came across this website where it explains that when you do batch processing, at least, flash attention is slower unless you concatenate or remove padding tokens or something...A little above my head but worth checking into if you're not seeing the benefit.

https://huggingface.co/docs/transformers/v4.40.2/en/perf_infer_gpu_one

![image](https://github.com/SYSTRAN/faster-whisper/assets/108230321/73961ffa-0af0-4f54-9413-d1d89e9a7aa2)

**Comment by AvivSham** [comment]
@BBC-Esq thanks! I'm using a single sample for inference, I just wonder if the lack of performance is fundamental or origin in the implementation.

@Purfview @Napuh @trungkienbkhn Could

**Comment by trungkienbkhn** [comment]
For information, I have executed [benchmark](https://github.com/SYSTRAN/faster-whisper/tree/master/benchmark) for FlashAttention with GPU NVIDIA H100 and large-v3 model.
Below are the results:

**1. Speed benchmark:**
Processing audio with duration 13:19.231s
Detected language 'fr' with probability 1.00
 
| System                           | Beam_size=5 | Beam_size=1 |
|-------------------------------- |-------------------- |-------------------- |
| Faster-Whisper              | 34.512s           | 27.190s           |
| FW with FlashAttention  | 33.751s           | 26.607s           |

**2. WER benchmark:**
Dataset: [librispeech_asr](https://huggingface.co/datasets/librispeech_asr)
Number of audio used for evaluation: 500
 
| System                           | Beam_size=5 | Beam_size=1 |
|---------------------------------|--------------------|--------------------|
| Faster-Whisper              | 2.649              | 2.325              |
| FW with FlashAttention  | 2.774             | 2.252             |

**3. Memory benchmark:**
GPU name: NVIDIA H100 PCIe
GPU device index: 0

| System                |Maximum increase of RAM|Maximum GPU memory usage| Maximum GPU power usage|
|------------------------|------------------------------------|-----------------------------------------|---------------------------------------|
| Faster-Whisper (beam_size=5)              | 1257 MiB  | 5178MiB / 81559MiB    | 157W / 350W                        |
| FW with FlashAttention (beam_size=5)  | 1251 MiB | 4954MiB / 81559MiB     | 153W / 350W                       |
| Faster-Whisper (beam_size=1)              | 1243 MiB  | 4634MiB / 81559MiB    | 164W / 350W                       |
| FW with FlashAttention (beam_size=1)  | 1327 MiB | 4602MiB / 81559MiB     | 164W / 350W                       |

=> Speed has improved a bit with FlashAttention
Note that currently, FlashAttention only [support](https://github.com/Dao-AILab/flash-attention?tab=readme-ov-file#installation-and-features) for Ampere, Ada, or Hopper GPUs (e.g., A100, RTX 3090, RTX 4090, H100)

**Comment by BBC-Esq** [comment]
Correct me if I'm wrong, but all of those results seem within the margin of error.  Just to confirm, you're running ctranslate 4.3?  4.3.1 just came out but don't think it'd make a difference.

I need to test on chat models first and see if there's a benefit there.  There was a noticeable improvement with beam size, but the ctranslate2 maintainer said he thought there'd be a larger benefit with a larger context length.  If there's a noticeable improvement with a large context length (while still keeping the beam size related improvements of course), then that means that the issues is specific to whisper.  The question then becomes: 1) is it an issue with the whisper model itself (e.g. the 30 second window), 2) an issue with ctranslate2's implementation of FA regarding whisper models, or 3) faster-whisper's implementation of ctranslate2...and perhaps something got lost in translation.  no pun intended.  ;-)

**Comment by BBC-Esq** [comment]
When I tested on chat models there was a noticeable improvement for all models EXCEPT, bizarely, llama2-based models.  And the improvement only manifested itself as follows:  When I increased the beam size, it was a little faster, but the VRAM increases was noticeably less than when FA is False...just a little backstory.

**Comment by AvivSham** [comment]
@trungkienbkhn Thank you for running this benchmark! 
Since we see marginal improvement (which can be explained by environment inconsistency), I wonder if this is the expected behavior or if there is an implementation issue.

**Comment by trungkienbkhn** [comment]
> Correct me if I'm wrong, but all of those results seem within the margin of error. Just to confirm, you're running ctranslate 4.3? 4.3.1 just came out but don't think it'd make a difference.

I executed with ctranslate2 4.2.1. But after discussing with @minhthuc2502, he confirmed that there is no difference when testing fw between 4.2.1 and 4.3.0.

> Since we see marginal improvement (which can be explained by environment inconsistency), I wonder if this is the expected behavior or if there is an implementation issue.

Yes it's expected behavior. There has been improvement, but not much. I think it might be because the number of input tokens is not large enough (maximum only 30 seconds per segment) and the fw large-v3 model is not too large in size (3GB).

**Comment by BBC-Esq** [comment]
I'll test it with the Phi3-128k and really push the context, see what benefit there is if any.  Run multiple benchmarks with and without FA to see.  To be clear, the VRAM savings wasn't inconsequential.  With different GPUs at differerent tiers (6,8, 10, 12, etc.) it really could be meaningful to someone.  Tokens/s, not so much...unless, again, it really shines with the 100k+ context or what not.

**Comment by BBC-Esq** [comment]
I tried to test Phi3-128k to test whether FA works for extremely long prompt lengths, however, I couldn't get it to run correctly.  Does anyone have a conversion of the Phi3-mini-128k or the Phi3-small-8k models by chance?  It seems like I can convert the 128k mini phi3 even though it still throws an error when I try to run inference on it, but the small-8k version I can't even convert.  Here's my conversion script for anyone's enjoyment.  Please let me know.
[convert_ctranslate2.txt](https://github.com/user-attachments/files/15845362/convert_ctranslate2.txt)

**Comment by trungkienbkhn** [comment]
> I tried to test Phi3-128k to test whether FA works for extremely long prompt lengths, however, I couldn't get it to run correctly. Does anyone have a conversion of the Phi3-mini-128k or the Phi3-small-8k models by chance? It seems like I can convert the 128k mini phi3 even though it still throws an error when I try to run inference on it, but the small-8k version I can't even convert. Here's my conversion script for anyone's enjoyment. Please let me know. [convert_ctranslate2.txt](https://github.com/user-attachments/files/15845362/convert_ctranslate2.txt)
> 
> Just FYI, converting these models requires ctranslate2==4.3+

You should post this question in [issues](https://github.com/OpenNMT/CTranslate2/issues) of ctranslate2 repo

**Comment by davidan208** [comment]
Hi guys,

I'm currently trying to use whisper with ct2 and flash attention as @trungkienbkhn's response.
However, I always get this line "Flash attention 2 is not supported" when trying to inference some samples.
Here is my environment: 
- A6000, CUDA 12.3, cuDNN 9.0
- Flash attention version 2.7.0.post2 (after using the default install line).

And these are my steps to run inference:
- Load whisper model using huggingface
- Convert to ct2 with this line
`ct2-transformers-converter --model models/whisper-large-v3-turbo --output_dir custom-faster-whisper-large-v3-turbo \
    --copy_files tokenizer.json preprocessor_config.json --quantization float16`
- Finally I use
`from faster_whisper import WhisperModel`
`model = WhisperModel('./models/whisper-large-v3-turbo', device="cuda", compute_type = 'float16', flash_attention=True)`

What could be the things that I had done incorrectly ? Please help !!!
Thank you in advance <3

**Comment by virtualmartire** [comment]
> Hi guys,
> 
> I'm currently trying to use whisper with ct2 and flash attention as @trungkienbkhn's response. However, I always get this line "Flash attention 2 is not supported" when trying to inference some samples. Here is my environment:
> 
> * A6000, CUDA 12.3, cuDNN 9.0, Python 3.10
> * Flash attention version 2.7.0.post2 (after using the default install line).
> 
> And these are my steps to run inference:
> 
> * Load whisper model using huggingface
> * Convert to ct2 with this line
>   `ct2-transformers-converter --model models/whisper-large-v3-turbo --output_dir custom-faster-whisper-large-v3-turbo \ --copy_files tokenizer.json preprocessor_config.json --quantization float16`
> * Finally I use
>   `from faster_whisper import WhisperModel`
>   `model = WhisperModel('./models/whisper-large-v3-turbo', device="cuda", compute_type = 'float16', flash_attention=True)`
> 
> What could be the things that I had done incorrectly ? Please help !!! Thank you in advance <3

same problem here
